### PR TITLE
#487: Refactor Description Bullets

### LINKED ISSUE #487: Backend Refactoring - Description Bullets
### Description

Refactor everything description bullets related

### Proposed Solution

1. separate each controller function into services and turn each into a class (see risk controller and risk service)
2. change all error calls to use the new custom exceptions (see error utils file and risk services to know what to throw)
3. move the query args to its own file (see risk query args file)
4. move the transformer to its own file (see risk transformer file)
5. update any existing tests to test the service function instead of using supertest and stuff (see risk test file)
6. update any testing types to use the prisma type and shared type for clarity (see risk test data file)
7. if any of the routes are using userId, delete it and use the auth userId instead (see editRisk controller)

### Info

see this epic and the attached pr for examples: https://github.com/Northeastern-Electric-Racing/FinishLine/issues/484

(pr: https://github.com/Northeastern-Electric-Racing/FinishLine/pull/446)

### PR DESCRIPTION
**Pull request by ryanhowe28**
## Changes

Separated controller function into service and turn into a class.
threw new exceptions accordingly
moved transformer and args to new files

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `yarn.lock` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #487

### CODE REVIEW

File: src/backend/src/prisma-query-args/description-bullets.query-args.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import { Prisma } from '@prisma/client';
+
+export const descBulletArgs = Prisma.validator<Prisma.Description_BulletArgs>()({
```
Comment (anthonybernardi) [inline src/backend/src/prisma-query-args/description-bullets.query-args.ts:3]: should be exported as the default like in risks

and lets rename it to `descriptionBulletQueryArgs`

File: src/backend/src/controllers/description-bullets.controllers.ts [hunk 2]
```diff
@@ -1,64 +1,17 @@
...
-    where: { descriptionId },
-    include: {
-      workPackageDeliverables: { include: { wbsElement: true } },
-      workPackageExpectedActivities: { include: { wbsElement: true } }
+import { NextFunction, Request, Response } from 'express';
+import DBService from '../services/description-bullets.services';
+import { getCurrentUser } from '../utils/utils';
+
+
+export default class DescriptionBulletController {
```
Comment (anthonybernardi) [inline src/backend/src/controllers/description-bullets.controllers.ts:6]: `DescriptionBulletsController`

File: src/backend/src/services/description-bullets.services.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
+import { User, WBS_Element_Status } from '@prisma/client';
+import prisma from '../prisma/prisma';
+import { hasBulletCheckingPermissions } from '../utils/description-bullets.utils';
+import { AccessDeniedException, HttpException, NotFoundException } from '../utils/errors.utils';
+import { descBulletTransformer } from '../transformers/description-bullets.transformer';
+import { descBulletArgs } from '../prisma-query-args/description-bullets.query-args';
+
+export default class DBService {
```
Comment (anthonybernardi) [inline src/backend/src/services/description-bullets.services.ts:8]: `DescriptionBulletsService`

File: src/backend/src/services/description-bullets.services.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  static async checkDescriptionBullet(userId: number, descriptionId: number, user: User) {
```
Comment (anthonybernardi) [inline src/backend/src/services/description-bullets.services.ts:9]: this function should have extensive documentation like in risks

File: src/backend/src/services/description-bullets.services.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    const workPackage = originalDB.workPackageDeliverables || originalDB.workPackageExpectedActivities;
+
+    if (!workPackage)
+      throw new HttpException(400, 'This description bullet is not tied to a workpackage deliverable or expected activity');
+
+    if (workPackage.wbsElement.status !== WBS_Element_Status.ACTIVE)
+      throw new HttpException(400, 'Cannot check a description bullet on an inactive work package!');
+
+    const hasPerms = await hasBulletCheckingPermissions(userId, descriptionId);
```
Comment (anthonybernardi) [inline src/backend/src/services/description-bullets.services.ts:29]: this should be checked usign user.userId

File: src/backend/src/services/description-bullets.services.ts (hunk 4 above)
Comment (anthonybernardi) [inline src/backend/src/services/description-bullets.services.ts:9]: we don't want userId at all anymore (even in the route). The way we get the user (`getCurrentUser`) is through auth and the way we got userId was just the frontend sending it which is not secure. so we can get rid of userId and also make user the first argument for consistency with other services

File: src/backend/src/routes/description-bullets.routes.ts [hunk 6]
```diff
@@ -10,7 +10,7 @@ descriptionBulletsRouter.post(
   body('userId').isInt({ min: 0 }).not().isString(),
```
Comment (anthonybernardi) [inline src/backend/src/routes/description-bullets.routes.ts:10]: dont need this line anymore, see my other comment

File: src/backend/src/transformers/description-bullets.transformer.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import { Prisma } from "@prisma/client";
+import { descBulletArgs } from "../prisma-query-args/description-bullets.query-args";
+
+
+export const descBulletTransformer = (descBullet: Prisma.Description_BulletGetPayload<typeof descBulletArgs>) => {
```
Comment (anthonybernardi) [inline src/backend/src/transformers/description-bullets.transformer.ts:5]: should export as default like risks

also lets rename it to `descriptionBulletTransformer`

File: src/backend/src/transformers/description-bullets.transformer.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+    return {
+      id: descBullet.descriptionId,
+      detail: descBullet.detail,
+      dateAdded: descBullet.dateAdded,
+      dateDeleted: descBullet.dateDeleted ?? undefined,
+      userChecked: descBullet.userChecked ?? undefined
+    };
+  };
```
Comment (anthonybernardi) [inline src/backend/src/transformers/description-bullets.transformer.ts:13]: missing newline at end of file, make sure you have prettier working on save

**Review by anthonybernardi — CHANGES_REQUESTED** [review]
nice work!

File: src/backend/src/services/description-bullets.services.ts (hunk 5 above)
Comment (ryanhowe28) [reply src/backend/src/services/description-bullets.services.ts]: might have made the change? not sure

**Comment by ryanhowe28** [comment]
don't know why all commits from develop showed up, moved to another branch/PR

**Pull request closed without merging**